src/lua/lua_tcp.c

```diff
@@ -1320,6 +1626,175 @@ lua_tcp_request (lua_State *L)
 	return 1;
 }
 
+/***
+ * @function rspamd_tcp.connect_sync({params})
+ * Creates new pseudo-synchronous connection to the specific address:port
+ *
+ * - `task`: rspamd task objects (implies `pool`, `session`, `ev_base` and `resolver` arguments)
+ * - `ev_base`: event base (if no task specified)
+ * - `resolver`: DNS resolver (no task)
+ * - `session`: events session (no task)
+ * - `config`: config (no task)
+ * - `host`: IP or name of the peer (required)
+ * - `port`: remote port to use
+ * - `timeout`: floating point value that specifies timeout for IO operations in **seconds**
+ * @return {boolean} true if request has been sent
+ */
+static gint
+lua_tcp_connect_sync (lua_State *L)
+{
+	LUA_TRACE_POINT;
+	GError *err = NULL;
+
+	gint64 port = -1;
+	gdouble timeout = default_tcp_timeout;
+	const gchar *host = NULL;
+	gint ret;
+	guint64 h;
+
+	struct rspamd_task *task = NULL;
+	struct rspamd_async_session *session = NULL;
+	struct rspamd_dns_resolver *resolver = NULL;
+	struct rspamd_config *cfg = NULL;
+	struct event_base *ev_base = NULL;
+
+	int arguments_validated = rspamd_lua_parse_table_arguments (L, 1, &err,
+			"task=U{task};session=U{session};resolver=U{resolver};ev_base=U{ev_base};"
+			"*host=S;*port=I;timeout=N;config=U{config}",
+			&task, &session, &resolver, &ev_base,
+			&host, &port, &timeout, &cfg);
+
+	if (!arguments_validated) {
+		if (err) {
+			ret = luaL_error (L, "invalid arguments: %s", err->message);
+			g_error_free (err);
+
+			return ret;
+		}
+
+		return luaL_error (L, "invalid arguments");
+	}
+
+	if (0 > port || port > 65535) {
+		return luaL_error (L, "invalid port given (correct values: 1..65535)");
+	}
+
+	if (task == NULL && (cfg == NULL || ev_base == NULL || session == NULL)) {
+		return luaL_error (L, "invalid arguments: either task or config+ev_base+session should be set");
+	}
+
+	if (timeout < 0.00001) {
+		/* rspamd_lua_parse_table_arguments() sets missing N field to zero */
+		timeout = default_tcp_timeout;
+	}
+	else {
+		timeout *= 1000.;
+	}
+	if (task) {
+		cfg = task->cfg;
+		ev_base = task->ev_base;
+		session = task->s;
+	}
+	if (resolver == NULL) {
+		if (task) {
+			resolver = task->resolver;
+		}
+		else {
+			resolver = lua_tcp_global_resolver (ev_base, cfg);
+		}
+	}
+
+	struct lua_tcp_cbdata *cbd = g_new0 (struct lua_tcp_cbdata, 1);
+
+	cbd->task = task;
+	cbd->cfg = cfg;
+	cbd->thread = lua_thread_pool_get_running_entry (cfg->lua_thread_pool);
+
+	h = rspamd_random_uint64_fast ();
+	rspamd_snprintf (cbd->tag, sizeof (cbd->tag), "%uxL", h);
+	cbd->handlers = g_queue_new ();
+
+	cbd->ev_base = ev_base;
+	cbd->flags |= LUA_TCP_FLAG_SYNC;
+	msec_to_tv (timeout, &cbd->tv);
```

Comment:
Why `msec_to_tv` and *= by 1000 previously? Perhaps, it is better to use `double_to_tv` and seconds here.